### PR TITLE
session: fix linearizability for non-autocommit async-commit txn

### PR DESCRIPTION
**Pull request by sticnarf**
### What problem does this PR solve?

https://github.com/pingcap/tidb/pull/22746 made a mistake. `s.GetSessionVars().IsAutocommit()` cannot be used to decide whether the current transaction is an auto-commit transaction.

Elsewhere we tend to use `s.GetSessionVars().IsAutocommit() && s.GetSessionVars().IsAutocommit().InTxn()` to check, but at the beginning of the `COMMIT` statement, we clear the `ServerStatusInTrans` flag, so `InTxn()` always returns false during the `COMMIT` statement.

### What is changed and how it works?

This PR adds `IsExplicit` flag to `txnCtx`. We set the flag to `true` when we start a transaction explicitly. We use this flag to check if this transaction is auto-commit.

### Check List 

Tests 

- Unit test

### Release note 

- No release note.

### CODE REVIEW

**Comment by sticnarf** [comment]
The way to fix may depend on https://github.com/pingcap/tidb/pull/22875

File: executor/simple.go [hunk 1]
```diff
@@ -586,7 +586,8 @@ func (e *SimpleExec) executeBegin(ctx context.Context, s *ast.BeginStmt) error {
 	// With START TRANSACTION, autocommit remains disabled until you end
 	// the transaction with COMMIT or ROLLBACK. The autocommit mode then
 	// reverts to its previous state.
-	e.ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusInTrans, true)
+	e.ctx.GetSessionVars().SetInTxn(true)
+	txnCtx.IsExplicit = true
```
Comment (AilinKid) [inline executor/simple.go:590]: why do the assignment again?

Comment (sticnarf) [reply executor/simple.go]: Sorry, I forgot to remove it during refactoring

**Review by AilinKid — COMMENTED** [review]
LGTM

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -164,9 +164,11 @@ type TransactionContext struct {
 	IsPessimistic  bool
 	// IsStaleness indicates whether the txn is read only staleness txn.
 	IsStaleness bool
-	Isolation   string
-	LockExpire  uint32
-	ForUpdate   uint32
+	// IsExplicit indicates whether the txn is an interactive txn started with a BEGIN statement
```
Comment (youjiali1995) [inline sessionctx/variable/session.go:167]: It's not accurate. A transaction can start with `autocommit = 0`.

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -1171,6 +1173,15 @@ func (s *SessionVars) GetStatusFlag(flag uint16) bool {
 	return s.Status&flag > 0
 }
 
+// SetInTxn sets whether the session is in transaction.
+// It also updates the IsExplicit flag in TxnCtx if val is true.
+func (s *SessionVars) SetInTxn(val bool) {
+	s.SetStatusFlag(mysql.ServerStatusInTrans, val)
+	if val {
+		s.TxnCtx.IsExplicit = true
```
Comment (youjiali1995) [inline sessionctx/variable/session.go:1181]: Is it reset when startinging a new transaction?

**Review by youjiali1995 — COMMENTED** [review]
/lgtm

File: sessionctx/variable/session.go (hunk 3 above)
Comment (sticnarf) [reply sessionctx/variable/session.go:1181]: I think it is always reset by `PrepareTxnCtx` or `NewTxn`. I am not sure if there is anything missing.

However, at least the unit test in this PR passes, which I think covers the important cases.

**Review by cfzjywxk — APPROVED** [review]
/lgtm

**Pull request merged**